### ISSUE TITLE
Errors when forwarding requests

### ISSUE THREAD

**Issue by xenon-ua** [jordan-breton/uws-reverse-proxy#2]
I am getting an error  `Error: socket hang up` in the console when make single reqest from browser (but the redirect is successful) and many errors `Gateway abortion` when I test proxy with autocannon

1. In first I added `console.error()` in the method #tryToRespondToError():
 ```javascript
#tryToRespondToError(error, uwsResponse, request){
  console.error('ERROR:', error)
  if(!this.#tryToUseErrorHandlerResponse(uwsResponse, error, request)){
    this.#tryToSendErrorResponse(uwsResponse, this.#buildErrorResponse(error));
  }
}
``` 
2. In second I created uws http server and proxy
 ```javascript
const port = 3005
const httpPort = 35794
const httpHost = '127.0.0.1'

// region Remote HTTP Server
const uWebSockets = require('uWebSockets.js')

uWebSockets.App()
  .get('/ping', res => {
    res.writeStatus('200 OK').end('PONG')
  })
  .listen(httpPort, listenSocket => {
    if(listenSocket) {
      console.log(`uWebSockets.js server started on http://localhost:${httpPort}`)
      return
    }
    console.error(`Unable to listen on ${httpPort} port`)
    process.exit(-1)
  })

// endregion Remote HTTP Server

// region Proxy server

const { UWSProxy, createUWSConfig, createHTTPConfig } = require('uws-reverse-proxy')

const proxy = new UWSProxy(
  createUWSConfig(uWebSockets, { port }),
  createHTTPConfig({ protocol: 'http', host: httpHost, port: httpPort })
)
proxy.start()
proxy.uws.server.listen('0.0.0.0', port, listening => {
  if(listening){
    console.log(`uWebSockets.js listening on port 0.0.0.0:${port}`);
  }else{
    console.error(`Unable to listen on port 0.0.0.0:${port}!`);
  }
})

// endregion
``` 
3. Start servers and make request `curl --location 'http://localhost:3005/ping'`. An error `Error: socket hang up`  will appear in the console
4. Run autocannon `autocannon -w 5 http://127.0.0.1:3005/ping`. This will give rise to many errors `Gateway abortion`.

Also using this proxy gives very low performance.
My results for proxy ping `autocannon -w 5 http://127.0.0.1:3005/ping`:
```
Avg Req/Sec: 1433
Avg Bytes/Sec: 238 kB
Avg Latency: 6,47 ms
```
My results for direct ping `autocannon -w 5 http://127.0.0.1:35794/ping`:
```
Avg Req/Sec: 19017
Avg Bytes/Sec: 1.83 MB
Avg Latency: 0.06 ms
```
This is probably because the Node.js `http` module is used to forward requests...

**Comment by jordan-breton** [comment]
Thanks for your detailed report 👍

### Benchmarks

Regarding benchmarks, it's expected to get a very bad result if you test against a uWebSocket.js http server, since no nodejs code can rivalize against it by a very huge factor.

Since the goal of using this proxy is especially to not use uWebSocket.js as an HTTP server but to forward requests to another server, it doesn't makes any sense to try to compare to it. By testing against an express/native/fastify http server, the gap is way smaller, but still unacceptable. It's something I'm aware of since I ran some benchmarks yesterday on my own, I must admit I didn't expect the `http` module to be that slow.

I started by enabling keep-alive for `http` and `https` agents, you should see an improvement, but it' still very bad.

### A solution to the `http` problem

Performance-wise, the `http` module is a real bottleneck. An option would be to use `undici` instead. After a first glance at the `undici` documentation, there is a stream API that would be very convenient to get responses as streams, but I didn't see any way to pass an http body as stream to send the request. I'll dig a bit further about it when I have some more time. 

Another solution would be to  use raw tcp sockets with `net` and `tls` modules.  I would be curious to compare both to see what is the most efficient, but it implies to also manage a connection pool. It's not trivial. But it may worth it, especially if I can keep an empty dependency graphe.

If you're aware of a more efficient http client, I'm open to any suggestion.

### Regarding errors

####  `Error: socket hang up`

I can't reproduce this one. My guess is that the `dev` branch fixed that somehow. It seems to be a not properly closed socket that was canceled by the http server or the curl command... Can you try to checkout to the dev branch to test it in your environment?

####  `Gateway Abortion`

I renamed this error for the next release as it was misleading. It happens when the server the proxy forward the request to abort the request. Since autocannon is squeezing the Proxy, it struggles to manage its connections so the http server abort the requests. It may also happen when the client (autocannon) abort the request because the proxy is unresponsive.

This one is a normal behavior.

**Comment by xenon-ua** [comment]
Thank you very much for the detailed investigation. Unfortunately, I cannot help with further research at the moment as I am away on a business trip and will not return for another week. However, I have a couple of comments.
Firstly, regarding the performance drop, the loss in performance is not 50%, but more than 10 times (19,017 -> 1,433 req/sec). Such a huge difference is still incomprehensible to me. Your idea of redirecting the request to a separate thread looks interesting, but it would need to be tested in practice. I don't think it will lead to a significant increase in performance.
If you created a chart based on my example, I think it would be more informative to label it as "Proxy Server (uWebSockets.js)" and "HTTP Server (uWebSockets.js)" to emphasize that both servers are built on uWebSockets.js and should not be a bottleneck causing performance degradation when handling requests.
Regarding the `Error: socket hang up`, I will try to reproduce it in the dev environment as soon as possible. The `Gateway Abortion` error does not seem normal to me, as I have not observed any request abortion from autocannon even on slow express servers that process only 600-700 requests per second. How will it looks for real clients during high load on the proxy server: they will send a request and the socket connection will be terminated with an error?

**Comment by jordan-breton** [comment]
# Error handling

> The Gateway Abortion error does not seem normal to me, as I have not observed any request abortion from autocannon 
> even on slow express servers that process only 600-700 requests per second.

There you have three components that can abort the request, so we have **3 points of failure**: the client, uWebSocket.js that handle the request and the targeted server. **Any** of those three components may fail for whatever reason (network, timeout, request aborted, etc.) 

In your test, when the load is **blocking/slowing down** the main thread, some requests can't be handled at all and are hanging somewhere between those three components, so at **any point in time** an IDLE connection (IDLE because the only thread available is struggling with all the work it has to do) can be closed either by the TCP stack or NodeJS internals or uWebSockets.js internals.

I think that it's what you've experienced: by being on a huge load, the thread struggle, some connections/requests are left IDLE and so they're closed/aborted.

>  How will it looks for real clients during high load on the proxy server: they will send a request and the socket connection will > be terminated with an error?

Yes, it's how it will work. The next release introduce a way to handle/customize those error responses. All of them are using the standard `502 Bad Gateway`, `503 Service Unavailable` and `504 Gateway Timeout` status codes by default and send a text/plain error message. Note that if this is the **client** itself or **uWebSockets.js** that close the connection, obviously no error code can be returned to the client.

# Performances

So, I finally found the time to do proper testing, because we really need to know what we're talking about. Sorry for what comes next, it will be a huge message.

> Firstly, regarding the performance drop, the loss in performance is not 50%, but more than 10 times (19,017 -> 1,433 req/sec). Such a huge difference is still incomprehensible to me. 

I didn't measure the 50% from the test you've done because **no NodeJS based HTTP server nor client** can compete with `uWebSockets.js` performances.

As I said earlier, trying to compete against `uWebSockets.js` doesn't make any sense, since the goal of `uws-reverse-proxy` is **to avoid** using `uWebSocket.js` as an HTTP server. What you tested is this:

```
uWebSockets.js -> proxy -> uWebSockets.js
```

In this configuration you don't need any proxy.

Since `uws-reverse-proxy` has not been written in C++, we can't expect to approach `uWebSocekts.js` performances, but that's not a problem since the target HTTP servers `uws-reverse-proxy` is intended to work with are **all** slower than `uWebSockets.js`. 

The goal for `uws-reverse-proxy` is to **not be** the **slowest** component of the stack. It must **only** be faster than the targeted HTTP servers (express, node:http itself, fastify, nestjs, koa, etc.).

The thing is that, actually, it is the **slowest** component in the stack, so its performances are slowing down the targeted HTTP server.

So when I mentioned an impact of 50%, I measured it by running a series of tests against the HTTP Client, since I assumed that it was the limiting factor. The idea is to use the **slowest** component in the stack to measure the overhead introduced by **uws-reverse-proxy** itself, first. 

Then, the goal is to see if we can reduce this overhead **and** if we can replace the slowest component in the stack by a better one.

So you'll find below all the tests I performed and my conclusion.

## Introduction

The first thing we must do is knowing which part of the proxy is limiting the overall performances.

Three components are involved:
- The receiver of the request: `uWebSocket.js`
- The HTTP client used to forward the request received by `uWebSocket.js` to the target server.
- The target server itself

My guess before running any test: the HTTP client is the limiting factor.

We will start by testing each component in isolation, then we will test the proxy and compare it to
an `Apache` proxy to see what would be an acceptable req/s rate to target. This rate will be the threshold
to determine if I can consider this performance issue as resolved when I implement a solution.

The test environment:

| Component | Specifications                                 |
| --------- | ---------------------------------------------- |
| OS        | Ubuntu 22.04.2 LTS 64 bits                     |
| Node      | v18.15.0                                       |
| RAM       | 16 Gio                                         |
| Processor | 11th Gen Intel® Core™ i7-11800H @ 2.30GHz × 16 |

No DNS resolution, all servers will listen on the loopback address.

All requests will only respond `'Hello world!'` with status `200 OK`

## `uWebSockets.js`

The file `/tests/benchmarks/isolation.uws.js` contains:

```js
const uWebSockets = require('uWebSockets.js');

const app = uWebSockets.App();

app.get('/', res => {
	res.cork(() => {
		res.end('Hello world!');
	});
});

app.listen('127.0.0.1', 3000, listenSocket => {
	if(listenSocket) {
		console.log('Listening to port 3000');
	} else {
		console.log('Failed to listen to port 3000');
	}
});
```

Test results:

```
Running 10s test @ http://127.0.0.1:3000/
10 connections
5 workers

|
┌─────────┬──────┬──────┬───────┬──────┬─────────┬─────────┬──────┐
│ Stat    │ 2.5% │ 50%  │ 97.5% │ 99%  │ Avg     │ Stdev   │ Max  │
├─────────┼──────┼──────┼───────┼──────┼─────────┼─────────┼──────┤
│ Latency │ 0 ms │ 0 ms │ 0 ms  │ 0 ms │ 0.01 ms │ 0.02 ms │ 6 ms │
└─────────┴──────┴──────┴───────┴──────┴─────────┴─────────┴──────┘
┌───────────┬─────────┬─────────┬─────────┬─────────┬─────────┬──────────┬─────────┐
│ Stat      │ 1%      │ 2.5%    │ 50%     │ 97.5%   │ Avg     │ Stdev    │ Min     │
├───────────┼─────────┼─────────┼─────────┼─────────┼─────────┼──────────┼─────────┤
│ Req/Sec   │ 210943  │ 210943  │ 255359  │ 262655  │ 249120  │ 14444.77 │ 210934  │
├───────────┼─────────┼─────────┼─────────┼─────────┼─────────┼──────────┼─────────┤
│ Bytes/Sec │ 22.2 MB │ 22.2 MB │ 26.8 MB │ 27.6 MB │ 26.2 MB │ 1.52 MB  │ 22.1 MB │
└───────────┴─────────┴─────────┴─────────┴─────────┴─────────┴──────────┴─────────┘

Req/Bytes counts sampled once per second.
# of samples: 54

2746k requests in 10.01s, 288 MB read
```

**Average req/s**: 249 120

## HTTP server

The file `/tests/benchmarks/isolation.http.js` contains:

```js
const http = require('http');

const server = http.createServer((req, res) => {
	res.writeHead(200);
	res.end('Hello world!');
});

server.listen(3000, '127.0.0.1', () => {
	console.log('HTTP server started on http://127.0.0.1:3000');
});
```

Test results:

```
Running 10s test @ http://127.0.0.1:3000/
10 connections
5 workers

|
┌─────────┬──────┬──────┬───────┬──────┬─────────┬─────────┬──────┐
│ Stat    │ 2.5% │ 50%  │ 97.5% │ 99%  │ Avg     │ Stdev   │ Max  │
├─────────┼──────┼──────┼───────┼──────┼─────────┼─────────┼──────┤
│ Latency │ 0 ms │ 0 ms │ 0 ms  │ 0 ms │ 0.01 ms │ 0.02 ms │ 4 ms │
└─────────┴──────┴──────┴───────┴──────┴─────────┴─────────┴──────┘
┌───────────┬─────────┬─────────┬─────────┬─────────┬─────────┬─────────┬─────────┐
│ Stat      │ 1%      │ 2.5%    │ 50%     │ 97.5%   │ Avg     │ Stdev   │ Min     │
├───────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┤
│ Req/Sec   │ 74879   │ 74879   │ 86591   │ 88447   │ 84595.2 │ 4685.49 │ 74847   │
├───────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┤
│ Bytes/Sec │ 11.5 MB │ 11.5 MB │ 13.2 MB │ 13.5 MB │ 12.9 MB │ 718 kB  │ 11.5 MB │
└───────────┴─────────┴─────────┴─────────┴─────────┴─────────┴─────────┴─────────┘

Req/Bytes counts sampled once per second.
# of samples: 54

932k requests in 10s, 143 MB read
```

**Average req/s**: 84 595

## HTTP Client

To test both HTTP clients, we will start an `uWebSockets.js` server with `node tests/benchmarks/isolation.uws.js`,
since it is the fastest HTTP server we have under our belt.

### `node:http`

Let's test the HTTP client. The file `/tests/benchmarks/isolation.http-client.js` contains:

```js
const http = require('http');
const { performance } = require('perf_hooks');

// The goal is to test in the same condition as autocannon
// Note that node:http does not support HTTP pipelining, it only support
// connection pooling + keep-alive
http.globalAgent = new http.Agent({ 
	keepAlive: true,
	maxSockets: 10 
});

// Define the number of requests to make
const NUM_REQUESTS = Number.parseInt(process.argv[2]) || 50000;
const START_TIME = performance.now();

let requestsEnded = 0;

for (let i = 0; i < NUM_REQUESTS; i++) {
	http.get('http://localhost:3000/', res => {
		res.on('data', () => {});
		res.on('close', () => {
			requestsEnded++;

			if(requestsEnded === NUM_REQUESTS){
				console.log(`Total requests: ${NUM_REQUESTS}`);
				console.log(`Total time taken: ${performance.now() - START_TIME}ms`);
				console.log(`Requests per second: ${NUM_REQUESTS / (performance.now() - START_TIME) * 1000}rps`);
			}
		});
	}).on('error', () => { console.error(); });
}
```

We start `uWebSockets.js` by running the command `node tests/benchmarks/isolation.uws.js` and then we run the HTTP client
by running the command `node tests/benchmarks/isolation.http-client.js 50000`.

Test results:

```
Total requests: 50000
Total time taken: 1917.7544709984213ms
Requests per second: 26071.197879130636rps
```

**Average req/s**: 26 071

### `autocannon`

`autocannon`, in another hand, is able to send much more requests per second while being written in JavaScript and
running with NodeJS too. Since they are comparable, let's see how `node:http` compete against it.

This time we'll use `autocannon` with the following command: `autocannon http://127.0.0.1:3000`, by default it will only use one thread and 10 connections.

There are the results: 

```
Running 10s test @ http://127.0.0.1:3000
10 connections
1 workers

|
┌─────────┬──────┬──────┬───────┬──────┬─────────┬─────────┬──────┐
│ Stat    │ 2.5% │ 50%  │ 97.5% │ 99%  │ Avg     │ Stdev   │ Max  │
├─────────┼──────┼──────┼───────┼──────┼─────────┼─────────┼──────┤
│ Latency │ 0 ms │ 0 ms │ 0 ms  │ 0 ms │ 0.01 ms │ 0.02 ms │ 4 ms │
└─────────┴──────┴──────┴───────┴──────┴─────────┴─────────┴──────┘
┌───────────┬─────────┬─────────┬─────────┬────────┬───────────┬─────────┬─────────┐
│ Stat      │ 1%      │ 2.5%    │ 50%     │ 97.5%  │ Avg       │ Stdev   │ Min     │
├───────────┼─────────┼─────────┼─────────┼────────┼───────────┼─────────┼─────────┤
│ Req/Sec   │ 105535  │ 105535  │ 112575  │ 114175 │ 112049.46 │ 2422.81 │ 105488  │
├───────────┼─────────┼─────────┼─────────┼────────┼───────────┼─────────┼─────────┤
│ Bytes/Sec │ 11.1 MB │ 11.1 MB │ 11.8 MB │ 12 MB  │ 11.8 MB   │ 255 kB  │ 11.1 MB │
└───────────┴─────────┴─────────┴─────────┴────────┴───────────┴─────────┴─────────┘

Req/Bytes counts sampled once per second.
# of samples: 11

1233k requests in 11.01s, 129 MB read

```

**Average req/s**: 112 049

## uws-reverse-proxy

Now that we have our base, let's see how much requests our proxy can handle by second, when the HTTP server is in another process.

Why?

Because we want our HTTP Client to be tested in **almost** the same situation than the `autocannon` HTTP client.
Since `autocannon` is running in its own process, the client used in `uws-reverse-proxy` must be in the least bloated
process possible. It must share the same thread as `uWebSockets.js` already, but it doesn't have to be impacted
by the HTTP server too, or we won't know what we will measure.

The file `/tests/benchmarks/proxy.separate-from-http-server.js` contains:

```js
const host = '127.0.0.1';
const port = 3005;
const httpPort = 3000;
const httpHost = '127.0.0.1';

const uWebSockets = require('uWebSockets.js');

// region Proxy server

const { UWSProxy, createUWSConfig, createHTTPConfig } = require('../../index');

const proxy = new UWSProxy(
	createUWSConfig(uWebSockets, { port }),
	createHTTPConfig({ protocol: 'http', host: httpHost, port: httpPort })
);
proxy.start();
proxy.uws.server.listen(host, port, listening => {
	if(listening){
		console.log(`uWebSockets.js listening on port ${host}:${port}`);
	}else{
		console.error(`Unable to listen on port ${host}:${port}!`);
	}
});

// endregion
```

Now let's start our proxy server:

```bash
node tests/benchmarks/proxy.separate-from-http-server.js
```

And our isolated HTTP server:

```bash
node tests/benchmarks/isolation.uws.js
```

And finally, let's test our proxy with `autocannon`:

```bash
autocannon http://127.0.0.1:3005
```

Here are the results:

```
Running 10s test @ http://127.0.0.1:3005
10 connections


┌─────────┬──────┬──────┬───────┬──────┬─────────┬─────────┬───────┐
│ Stat    │ 2.5% │ 50%  │ 97.5% │ 99%  │ Avg     │ Stdev   │ Max   │
├─────────┼──────┼──────┼───────┼──────┼─────────┼─────────┼───────┤
│ Latency │ 0 ms │ 0 ms │ 1 ms  │ 1 ms │ 0.03 ms │ 0.21 ms │ 17 ms │
└─────────┴──────┴──────┴───────┴──────┴─────────┴─────────┴───────┘
┌───────────┬─────────┬─────────┬─────────┬─────────┬─────────┬─────────┬─────────┐
│ Stat      │ 1%      │ 2.5%    │ 50%     │ 97.5%   │ Avg     │ Stdev   │ Min     │
├───────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┤
│ Req/Sec   │ 12535   │ 12535   │ 16911   │ 17439   │ 16357.6 │ 1422.59 │ 12532   │
├───────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┤
│ Bytes/Sec │ 2.18 MB │ 2.18 MB │ 2.94 MB │ 3.04 MB │ 2.85 MB │ 248 kB  │ 2.18 MB │
└───────────┴─────────┴─────────┴─────────┴─────────┴─────────┴─────────┴─────────┘

Req/Bytes counts sampled once per second.
# of samples: 10

164k requests in 10.01s, 28.5 MB read

```

**Average req/s**: 16 357

MMh. This is quite bad. That's **16 900** requests per second on average. It's 35% less than the `node:http` client.

The question is... how a well-known reverse proxy would perform in the same situation? Let's measure
it with `apache2`.

## `Apache`

We start by creating the following configuration file:

```apache title="/etc/apache2/conf-available/proxy.conf"
<VirtualHost *:80>
    ServerName localhost
    
    ProxyRequests Off
    ProxyPreserveHost On
    
    ProxyPass / http://127.0.0.1:3000/
    ProxyPassReverse / http://127.0.0.1:3000/
</VirtualHost>
```

We enable the configuration:

```bash
sudo a2enconf proxy
sudo systemctl reload apache2
```

We check it's alive:

```bash
curl http://127.0.0.1/
```

It successfully returns the `Hello World!` message. We can use `autocannon`: 

```bash
autocannon http://127.0.0.1/
```

And got the following results:

```
Running 10s test @ http://127.0.0.1
10 connections


┌─────────┬──────┬───────┬────────┬────────┬──────────┬──────────┬────────┐
│ Stat    │ 2.5% │ 50%   │ 97.5%  │ 99%    │ Avg      │ Stdev    │ Max    │
├─────────┼──────┼───────┼────────┼────────┼──────────┼──────────┼────────┤
│ Latency │ 1 ms │ 54 ms │ 111 ms │ 116 ms │ 54.66 ms │ 33.56 ms │ 132 ms │
└─────────┴──────┴───────┴────────┴────────┴──────────┴──────────┴────────┘
┌───────────┬─────────┬─────────┬─────────┬─────────┬──────────┬─────────┬─────────┐
│ Stat      │ 1%      │ 2.5%    │ 50%     │ 97.5%   │ Avg      │ Stdev   │ Min     │
├───────────┼─────────┼─────────┼─────────┼─────────┼──────────┼─────────┼─────────┤
│ Req/Sec   │ 69439   │ 69439   │ 73919   │ 83967   │ 74551.28 │ 4526.09 │ 69383   │
├───────────┼─────────┼─────────┼─────────┼─────────┼──────────┼─────────┼─────────┤
│ Bytes/Sec │ 13.3 MB │ 13.3 MB │ 14.2 MB │ 16.1 MB │ 14.3 MB  │ 867 kB  │ 13.3 MB │
└───────────┴─────────┴─────────┴─────────┴─────────┴──────────┴─────────┴─────────┘

Req/Bytes counts sampled once per second.
# of samples: 11

828k requests in 11.01s, 157 MB read
```

**Average req/s**: 74 551

## Conclusion

Ok, let's resume our tests so far:

| Test                             | Average req/s |
|----------------------------------|---------------|
| Standalone uWebSockets.js server | 255 000       |
| Standalone `node:http` server    | 84 500        |
| autocannon HTTP client           | 112 000       |
| Standalone `node:http` client    | 26 000        |
| Standalone uws-reverse-proxy     | 17 000        |
| Apache2 reverse proxy            | 74 500        |

What we can say about it:

1) `autocannon` in one worker can't overload `uWebSockets.js`
2) `uWebSockets.js` is the fastest component of the stack (not surprising)
3) `node:http` client is not suitable to use as an HTTP client in `uws-reverse-proxy` as we can see that it's almost 5 times slower than what `autocannon` can do with the same number of open connections.
4) A regular reverse proxy like `apache` is able to handle more than 4 times more requests than `uws-reverse-proxy`.

The most limiting factor here seems to be `node:http` client. On top of that, `uws-reverse-proxy` seems to eat
up to 35% of the capacity of the HTTP client, since we measured that `uws-reverse-proxy` handle 17 000 req/s while the maximum amount of req/s we may expect from a `node:http` client is 26 000 req/s.

## And now what?

By changing the HTTP client, we should be able to get somewhere around 72 000 req/s (autocannon 112 000 req/s - 35% of `uws-reverse-proxy` processing), that would be comparable to `apache2`. This seems to be a consistent estimation.

So the next step is to implement an HTTP client similar to the one used by autocannon and see what happens.

**Comment by jordan-breton** [comment]
In the `dev` branch, the HTTP client have been changed for a custom implementation. It can go up to 90 000 req/s using connection pooling and pipelining. 

It's very intresting, because proxy's performances got better, but still far below what I would have expected (17 000 -> 25 000 req/s). I'm still profiling, but I/Os and CPU usage seems to play a huge role in this loss. When inspecting flamegraphes, I also see that NodeJS streams buffer have a huge impact (_genericWrite).  

I don't yet said my last words. The problem seems even more exciting to solve, finally :)

**Comment by jordan-breton** [comment]
So, I finally published the [3.2.0 release](https://github.com/jordan-breton/uws-reverse-proxy/releases/tag/v3.2.0)

It brings a **300%** performances improvement to the proxy. I added a note about performances in the README to. ReadStreams and the native `http` module where huge bottlenecks. By using pipelining, I got something far more acceptable, but it adds some complexity.

I close this issue since I feels this improvement sufficient for most use cases. Feels free to investigate and submit a PR if you have an idea to push it further.
